// xdl/build/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bunyan_1 = __importDefault(require("@expo/bunyan"));
class ConsoleRawStream {
    write(record) {
        if (record.level < bunyan_1.default.INFO) {
            console.log(record);
        }
        else if (record.level < bunyan_1.default.WARN) {
            console.info(record);
        }
        else if (record.level < bunyan_1.default.ERROR) {
            console.warn(record);
        }
        else {
            console.error(record);
        }
    }
}
const logger = bunyan_1.default.createLogger({
    name: 'expo',
    serializers: bunyan_1.default.stdSerializers,
    streams: process.env.EXPO_RAW_LOG && process.env.NODE_ENV !== 'production'
        ? [
            {
                type: 'raw',
                stream: new ConsoleRawStream(),
                closeOnExit: false,
                level: 'debug',
            },
        ]
        : [],
});
exports.default = {
    child: (options) => logger.child(options),
    notifications: logger.child({ type: 'notifications' }),
    global: logger.child({ type: 'global' }),
    DEBUG: bunyan_1.default.DEBUG,
    INFO: bunyan_1.default.INFO,
    WARN: bunyan_1.default.WARN,
    ERROR: bunyan_1.default.ERROR,
};
