// xdl/build/UserSettings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const json_file_1 = __importDefault(require("@expo/json-file"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const uuid_1 = __importDefault(require("uuid"));
const internal_1 = require("./internal");
const SETTINGS_FILE_NAME = 'state.json';
function userSettingsFile() {
    const dir = dotExpoHomeDirectory();
    const file = path_1.default.join(dir, SETTINGS_FILE_NAME);
    try {
        // move exponent.json to state.json
        const oldFile = path_1.default.join(dir, 'exponent.json');
        if (fs_extra_1.default.statSync(oldFile).isFile()) {
            fs_extra_1.default.renameSync(oldFile, file);
        }
    }
    catch (e) {
        // no old directory, continue
    }
    return file;
}
function userSettingsJsonFile() {
    return new json_file_1.default(userSettingsFile(), {
        jsonParseErrorDefault: {},
        cantReadFileDefault: {},
    });
}
let mkdirped = false;
function dotExpoHomeDirectory() {
    let dirPath;
    if (process.env.__UNSAFE_EXPO_HOME_DIRECTORY) {
        dirPath = process.env.__UNSAFE_EXPO_HOME_DIRECTORY;
    }
    else {
        const home = internal_1.Env.home();
        if (!home) {
            throw new Error("Can't determine your home directory; make sure your $HOME environment variable is set.");
        }
        if (process.env.EXPO_STAGING) {
            dirPath = path_1.default.join(home, '.expo-staging');
        }
        else if (process.env.EXPO_LOCAL) {
            dirPath = path_1.default.join(home, '.expo-local');
        }
        else {
            dirPath = path_1.default.join(home, '.expo');
        }
        try {
            // move .exponent to .expo
            const oldDirPath = path_1.default.join(home, '.exponent');
            if (fs_extra_1.default.statSync(oldDirPath).isDirectory()) {
                fs_extra_1.default.renameSync(oldDirPath, dirPath);
            }
        }
        catch (e) {
            // no old directory, continue
        }
    }
    if (!mkdirped) {
        fs_extra_1.default.mkdirpSync(dirPath);
        mkdirped = true;
    }
    return dirPath;
}
// returns an anonymous, unique identifier for a user on the current computer
async function anonymousIdentifier() {
    const settings = await userSettingsJsonFile();
    let id = await settings.getAsync('uuid', null);
    if (!id) {
        id = uuid_1.default.v4();
        await settings.setAsync('uuid', id);
    }
    return id;
}
function accessToken() {
    return process.env.EXPO_TOKEN || null;
}
const UserSettings = Object.assign(userSettingsJsonFile(), {
    dotExpoHomeDirectory,
    userSettingsFile,
    userSettingsJsonFile,
    accessToken,
    anonymousIdentifier,
    SETTINGS_FILE_NAME,
});
exports.default = UserSettings;
